controllers/customer_controller.go

```diff
@@ -0,0 +1,221 @@
+	Scheme                *runtime.Scheme
+	JiraServiceDeskClient jiraservicedeskclient.Client
+}
+
+// +kubebuilder:rbac:groups=jiraservicedesk.stakater.com,resources=customers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=jiraservicedesk.stakater.com,resources=customers/status,verbs=get;update;patch
+
+func (r *CustomerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	log := r.Log.WithValues("customer", req.NamespacedName)
+
+	log.Info("Reconciling Customer")
+
+	// Fetch the Customer instance
+	instance := &jiraservicedeskv1alpha1.Customer{}
+
+	err := r.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcilerUtil.DoNotRequeue()
+		}
+		// Error reading the object - donot requeue the request
+		return reconcilerUtil.RequeueWithError(err)
+	}
+
+	// Validate Custom Resource
+	if ok, err := instance.IsValid(); !ok {
+		return reconcilerUtil.ManageError(r.Client, instance, err, false)
+	}
+
+	// Resource is marked for deletion
+	if instance.DeletionTimestamp != nil {
+		log.Info("Deletion timestamp found for instance " + req.Name)
+		if finalizerUtil.HasFinalizer(instance, CustomerFinalizer) {
+			return r.handleDelete(req, instance)
+		}
+		// Finalizer doesn't exist so clean up is already done
+		return reconcilerUtil.DoNotRequeue()
+	}
+
+	// Add finalizer if it doesn't exist
+	if !finalizerUtil.HasFinalizer(instance, CustomerFinalizer) {
+		log.Info("Adding finalizer for instance " + req.Name)
+
+		finalizerUtil.AddFinalizer(instance, CustomerFinalizer)
+
+		err := r.Client.Update(context.TODO(), instance)
+		if err != nil {
+			return reconcilerUtil.ManageError(r.Client, instance, err, false)
+		}
+
+		return reconcilerUtil.DoNotRequeue()
+	}
+
+	if len(instance.Status.CustomerId) > 0 {
+		existingCustomer, err := r.JiraServiceDeskClient.GetCustomerById(instance.Status.CustomerId)
+		if err != nil {
+			return reconcilerUtil.ManageError(r.Client, instance, err, false)
+		}
+		return r.handleUpdate(req, existingCustomer, instance)
+	}
+
+	return r.handleCreate(req, instance)
+}
+
+func (r *CustomerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&jiraservicedeskv1alpha1.Customer{}).
+		Complete(r)
+}
+
+func (r *CustomerReconciler) handleUpdate(req ctrl.Request, existingCustomer jiraservicedeskclient.Customer, instance *jiraservicedeskv1alpha1.Customer) (ctrl.Result, error) {
+	log := r.Log.WithValues("customer", req.NamespacedName)
+
+	log.Info("Modifying project associations for JSD Customer: " + instance.Spec.Name)
+
+	existingCustomerInstance := r.JiraServiceDeskClient.GetCustomerCRFromCustomer(existingCustomer)
```

Comment:
Do we really need this for update ?

Reply:
Shouldn't we just use instance for verification ?

Reply:
The existing customer is get for validation checking to make sure that the customer hasn't changed the customer name or email in the Custom Resource. In case a user change any of these fields an error is returned as they shouldn't be changed.

Reply:
Please use `instance` in this case instead of using the customer retrieved using API call. We should be relying on the custom resources for validating update/create and then rely on the get customer API call just for ensuring that the customer still exists.